### ISSUE TITLE
Rethink flags presets

### ISSUE THREAD

**Issue by sharadhr** [friendlyanon/cmake-init#94]
The various `flags-*` presets are partially repetitive, don't fully represent the different setups that people could possibly have, which makes extensibility difficult. For instance, adding different flags for different architectures, or using different compilers on Windows (`clang.exe` with the GNU-like command-line is available with MSVC itself).

I propose the following:

1. `flags-linux` and `flags-darwin` be merged and renamed to `gnu-compiler-flags` as a sort of 'base' preset, from which other (custom) presets may inherit. At present the only difference between the two is `-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3`, which can be handled separately, as noted below in (4).
2. Separate out the linker flags from the compiler flags preset into `gnu-linker-flags`. 
3. `flags-windows` be renamed to `cl-compiler-flags`. Once again, MSVC includes `clang-cl` which these can be used with, too. Windows also comes with MinGW, which includes `gcc` and `g++` *for Windows*, in which case `flags-linux` is really a misnomer. 
4. Consider using the `$env{VARIABLE}` trick detailed in [this StackOverflow answer](https://stackoverflow.com/a/70791991/1654223) to combine flags. I concede this is not elegant (pollutes the CMake build-time environment), but until CMake itself provides a solution to [this issue](https://gitlab.kitware.com/cmake/cmake/-/issues/22538), I don't see any other way around it. 

Cheers.

**Comment by friendlyanon** [comment]
I am aware that the flags presets are repetitive and that they don't cover every possible setup, which is mentioned in the non-goals of this project: https://github.com/friendlyanon/cmake-init#non-goals  
The idea behind the flags is that they are mainly for CI that also happens to work on "default" configurations with good enough experience. `$env{}` will definitely not be abused.

User presets can and **are meant to** also be freely used to override parts of the main presets. With that in mind, I'm not sure what to change in that regard. Regarding MinGW, that was looked at in https://github.com/friendlyanon/cmake-init/issues/89 and I don't see what needs to be done here. MinGW is definitely not the norm for Windows, so I would rather not have it in cmake-init by default.

Fortify not working with ASAN was no something I was aware of though. From what I read, it's best to just remove the `-D` one in this case.

**Comment by sharadhr** [comment]
> that they don't cover every possible setup, which is mentioned in the non-goals of this project: https://github.com/friendlyanon/cmake-init#non-goals

That's fair, and I don't think it's reasonable to expect `cmake-init` to support every possible combination of compiler and platform there is. However, I think renaming what is already present would make things clearer—the real difference between the `flags-linux` + `flags-darwin` and `flags-windows` sets are the syntax of the driver flags, rather than the OS they're running on. 

> `$env{}` will definitely not be abused

Bummer. This is a CMake limitation, honestly, and it does feel weird to require this sort of env-variable mangling...

> MinGW

Simply put, `flags-windows` can't be used to compile with MinGW, simply because the tools provided by MinGW don't accept `cl`-style flags (it *is* called Minimalist **GNU** for Windows). Ergo I said `flags-linux` is not particularly well-named: it doesn't really contain anything strictly Linux-specific, is in the GNU style of compiler arguments, and it *could* be set up **by end-users** to compile for *any* platform that uses GCC, Clang, ICX, NVCC, or any number of compilers that have a shared GNU-like heritage. 

Ergo this whole renaming exercise.

**Comment by friendlyanon** [comment]
I guess renaming them is a fair suggestion, the names were given with each platform's "default" configuration (Windows - MSVC, Linux - GCC/Clang, Mac - AppleClang) in mind, but putting those in the name instead would more clearly describe their intended usage to someone trying to use e.g. MinGW on Windows. As an FYI, some of the hardening flags are not applicable for AppleClang, that's why the *nix preset was split.

> it does feel weird to require this sort of env-variable mangling

You shouldn't do that. You can provide CMake with the flags using a toolchain file. Tooling will still understand the build perfectly fine if you instead load a toolchain file in the preset. Toolchain files are just CMake scripts, where you can put some sort of logic to better select flags that are currently in the presets, but I would heavily recommend keeping that logic very light. Remember that the whole point of the presets is to externalize things that aren't pertinent to build and usage requirements.
